### DISCUSSION TITLE
Future features Battery-Cycler

### DISCUSSION THREAD

**Discussion by pastorpflores** [WattRex/Regenerative-Battery-Cycler#90] [Ideas]
# Next ideas to improve design

- [ ] [Measurement sync](https://github.com/orgs/WattRex/discussions/90#discussioncomment-6280933)

**Comment by pastorpflores** [comment]
## Measurement:

ADC trigger can be attached to PWM mid-point, to ensure syncronization between middle-point of coil current curve and reduce noise. As, PWM signal has a frequency of 500 kHz and measures are done in a 100 kHz rate, a counter of 5 PWM mid-point irqs must be configured to trigger the ADC. 

Review [AN4539 Application Note HRTIM Cookbook](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwiS5uXc4uD_AhV3UKQEHdc8AdYQFnoECBAQAQ&url=https%3A%2F%2Fwww.st.com%2Fresource%2Fen%2Fapplication_note%2Fan4539-hrtim-cookbook-stmicroelectronics.pdf&usg=AOvVaw09GoSPuixRqPjGPV_Od9LM&opi=89978449). 

Something like that would be desirable:
![imagen](https://github.com/WattRex/Regenerative-Battery-Cycler/assets/48552041/a549131a-25f3-48d8-9867-97bc51916804)

**Comment by pastorpflores** [comment]
## Comm queue:
Manage CAN communications with an interal msg queue in MID_COMM. In each cycle read all msgs in the CAN queue and store them in an internal queue. Then, in each cycle APP_IFACE will process a single msg (or several) to improve cycle time.

**Comment by RobertoAldea** [comment]
Check the control mode and verify the limit is acceptable:

- Control Current Mode -> limits for Power, Voltage and Time.
- Control Voltage Mode -> limits for Power, Current and Time
- Control Power Mode -> limits for Current,  Voltage and Time.
- Wait Mode ->Limit for Time
